hal/targets/hal/TARGET_NORDIC/TARGET_NRF5/sleep.c

```diff
@@ -16,13 +16,57 @@
 #include "sleep_api.h"
 #include "cmsis.h"
 #include "mbed_interface.h"
+#include "nrf_soc.h"
+#include "nrf_sdm.h"
+
+// Mask of reserved bits of the register ICSR in the System Control Block peripheral
+// In this case, bits which are equal to 0 are the bits reserved in this register
+#define SCB_ICSR_RESERVED_BITS_MASK     0x9E43F03F
 
 void sleep(void)
 {
     // ensure debug is disconnected if semihost is enabled....
-    NRF_POWER->TASKS_LOWPWR = 1;
-    // wait for interrupt
-    __WFE();
+
+    SCB->SCR |= SCB_SCR_SEVONPEND_Msk; /* send an event when an interrupt is pending.
+                                        * This helps with the wakeup from the following app_evt_wait(). */
+
+    uint8_t sd_enabled;
+
+    // look if exceptions are enabled or not, if they are, it is possible to make an SVC call
+    // and check if the soft device is running
+    if ((__get_PRIMASK() == 0) && (sd_softdevice_is_enabled(&sd_enabled) == NRF_SUCCESS) && (sd_enabled == 1)) {
```

Comment:
Good point. I took the code from mbedOS 3 and only corrected it, so it would run on nRF52. But indeed, it should be refactored a bit more.